spec/lib/notification_report_spec.rb

```diff
@@ -0,0 +1,152 @@
+require "rails_helper"
+
+RSpec.describe NotificationReport do
+  Entry = NotificationReport::Entry
+  Printer = NotificationReport::Printer
+
+  subject { described_class.all }
+
+  it "has an entry per request id" do
+    create(:notification_log, govuk_request_id: "a")
+    create(:notification_log, govuk_request_id: "b")
+    create(:notification_log, govuk_request_id: "a")
+
+    expect(subject.entries.count).to eq(2)
+  end
+
+  describe Entry do
+    let(:entry) { Entry.new("request_id", NotificationLog.all) }
+
+    it "reports the notifications sent from each app" do
+      x = create(:notification_log, emailing_app: "email_alert_api")
+      y = create(:notification_log, emailing_app: "gov_uk_delivery")
+      z = create(:notification_log, emailing_app: "gov_uk_delivery")
+
+      expect(entry.email_alert_api_notifications).to match_array [x]
+      expect(entry.gov_uk_delivery_notifications).to match_array [y, z]
+    end
+
+    describe "#topics_matched_in_both_systems" do
+      it "reports the intersection of gov_delivery_ids" do
+        create(:notification_log, emailing_app: "email_alert_api", gov_delivery_ids: %w(a b c))
+        create(:notification_log, emailing_app: "gov_uk_delivery", gov_delivery_ids: %w(b c d))
+
+        expect(entry.topics_matched_in_both_systems).to match_array %w(b c)
+      end
+
+      it "returns an empty array if a notification wasn't sent from either app" do
+        create(:notification_log, emailing_app: "email_alert_api", gov_delivery_ids: %w(a b c))
+        expect(entry.topics_matched_in_both_systems).to be_empty
+      end
+
+      context "when multiple notifications were sent from a system" do
+        before do
+          create(:notification_log, emailing_app: "email_alert_api", gov_delivery_ids: %w(a b c))
+
+          create(:notification_log, emailing_app: "gov_uk_delivery", gov_delivery_ids: %w(b c d))
+          create(:notification_log, emailing_app: "gov_uk_delivery", gov_delivery_ids: %w(c d))
+        end
+
+        it "uses the gov_delivery_ids of the last notification log" do
```

Comment:
Ah, just read through the rest of the spec 😄  It looks like there's alerting in place for when the multiple notifications' gov_delivery_ids don't match (e.g. `#email_alert_api_notifications_have_the_same_topics`), so that would alert us to the fact that something fishy is going on, and we can then investigate. So you can ignore the previous comment!